// @ngx-loading-bar/router/fesm5/ngx-loading-bar-router.js
import { __decorate, __metadata } from 'tslib';
import { NgModule } from '@angular/core';
import { NavigationStart, NavigationError, NavigationEnd, NavigationCancel, Router, RouterModule } from '@angular/router';
import { LoadingBarService, LoadingBarModule } from '@ngx-loading-bar/core';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@ngx-loading-bar/core';
var LoadingBarRouterModule = /** @class */ (function () {
    function LoadingBarRouterModule(router, loader) {
        var _this = this;
        var ref = loader.useRef('router');
        router.events.subscribe(function (event) {
            var navState = _this.getCurrentNavigationState(router);
            if (navState && navState.ignoreLoadingBar) {
                return;
            }
            if (event instanceof NavigationStart) {
                ref.start();
            }
            if (event instanceof NavigationError || event instanceof NavigationEnd || event instanceof NavigationCancel) {
                ref.complete();
            }
        });
    }
    LoadingBarRouterModule.prototype.getCurrentNavigationState = function (router) {
        // `getCurrentNavigation` only available in angular `7.2`
        var currentNavigation = router.getCurrentNavigation && router.getCurrentNavigation();
        if (currentNavigation && currentNavigation.extras) {
            return currentNavigation.extras.state;
        }
        return {};
    };
    LoadingBarRouterModule.ctorParameters = function () { return [
        { type: Router },
        { type: LoadingBarService }
    ]; };
    LoadingBarRouterModule = __decorate([ __metadata("design:paramtypes", [Router, LoadingBarService])
    ], LoadingBarRouterModule);
LoadingBarRouterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LoadingBarRouterModule });
LoadingBarRouterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LoadingBarRouterModule_Factory(t) { return new (t || LoadingBarRouterModule)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(ɵngcc2.LoadingBarService)); }, imports: [[RouterModule, LoadingBarModule], RouterModule, LoadingBarModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LoadingBarRouterModule, { imports: function () { return [RouterModule, LoadingBarModule]; }, exports: function () { return [RouterModule, LoadingBarModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LoadingBarRouterModule, [{
        type: NgModule,
        args: [{
                imports: [RouterModule, LoadingBarModule],
                exports: [RouterModule, LoadingBarModule]
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.LoadingBarService }]; }, null); })();
    return LoadingBarRouterModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { LoadingBarRouterModule };
